File: go/formatter.go
package devlogs

import (
	"bytes"
	"context"
	"log/slog"
	"os"
	"runtime"
	"strconv"
	"strings"
	"time"
)

// LogSource contains source location info (v2.0 schema).
type LogSource struct {
	Logger   string  `json:"logger"`
	Pathname *string `json:"pathname"`
	LineNo   *int    `json:"lineno"`
	FuncName *string `json:"funcName"`
}

// LogProcess contains process/thread info (v2.0 schema).
type LogProcess struct {
	ID     int `json:"id"`
	Thread int `json:"thread"`
}

// LogDocument represents the document structure sent to OpenSearch (v2.0 schema).
type LogDocument struct {
	DocType string `json:"doc_type"`

	// Required fields
	Application string `json:"application"`
	Component   string `json:"component"`
	Timestamp   string `json:"timestamp"`

	// Top-level log fields
	Message string  `json:"message"`
	Level   string  `json:"level"`
	Area    *string `json:"area"`

	// Optional metadata
	Environment *string `json:"environment,omitempty"`
	Version     *string `json:"version,omitempty"`
	OperationID *string `json:"operation_id"`

	// Custom fields (renamed from features)
	Fields map[string]interface{} `json:"fields,omitempty"`

	// Source and process info
	Source    LogSource  `json:"source"`
	Process   LogProcess `json:"process"`
	Exception *string    `json:"exception,omitempty"`
}

// FormatLogDocument converts an slog.Record to a LogDocument using v2.0 schema.
func FormatLogDocument(ctx context.Context, r slog.Record, cfg *Config) *LogDocument {
	doc := &LogDocument{
		DocType:     "log_entry",
		Application: cfg.Application,
		Component:   cfg.Component,
		Timestamp:   r.Time.UTC().Format("2006-01-02T15:04:05.000Z"),
		Message:     r.Message,
		Level:       NormalizeLevel(r.Level),
		Source: LogSource{
			Logger: cfg.Component, // Use component as default logger name
		},
		Process: LogProcess{
			ID:     os.Getpid(),
			Thread: getGoroutineID(),
		},
	}

	// Optional metadata
	if cfg.Environment != "" {
		doc.Environment = &cfg.Environment
	}
	if cfg.Version != "" {
		doc.Version = &cfg.Version
	}

	// Extract source info
	if r.PC != 0 {
		frames := runtime.CallersFrames([]uintptr{r.PC})
		frame, _ := frames.Next()
		if frame.File != "" {
			doc.Source.Pathname = &frame.File
		}
		if frame.Line > 0 {
			doc.Source.LineNo = &frame.Line
		}
		if frame.Function != "" {
			doc.Source.FuncName = &frame.Function
			// Use function name as logger if available
			doc.Source.Logger = frame.Function
		}
	}

	// Get context values
	if area := GetArea(ctx); area != "" {
		doc.Area = &area
	}
	if opID := GetOperationID(ctx); opID != "" {
		doc.OperationID = &opID
	}

	// Extract fields from record attributes (renamed from features)
	fields := make(map[string]interface{})
	r.Attrs(func(a slog.Attr) bool {
		fields[a.Key] = resolveValue(a.Value)
		return true
	})
	if len(fields) > 0 {
		doc.Fields = fields
	}

	return doc
}

// resolveValue converts slog.Value to a JSON-serializable value.
func resolveValue(v slog.Value) interface{} {
	switch v.Kind() {
	case slog.KindString:
		return v.String()
	case slog.KindInt64:
		return v.Int64()
	case slog.KindUint64:
		return v.Uint64()
	case slog.KindFloat64:
		return v.Float64()
	case slog.KindBool:
		return v.Bool()
	case slog.KindDuration:
		return v.Duration().String()
	case slog.KindTime:
		return v.Time().Format(time.RFC3339)
	case slog.KindGroup:
		attrs := v.Group()
		m := make(map[string]interface{}, len(attrs))
		for _, a := range attrs {
			m[a.Key] = resolveValue(a.Value)
		}
		return m
	case slog.KindAny:
		return v.Any()
	default:
		return v.String()
	}
}

// getGoroutineID extracts the goroutine ID from runtime.Stack.
func getGoroutineID() int {
	var buf [64]byte
	n := runtime.Stack(buf[:], false)
	// Stack output starts with "goroutine <id> [...]"
	s := string(buf[:n])
	if strings.HasPrefix(s, "goroutine ") {
		s = s[len("goroutine "):]
		if idx := strings.IndexByte(s, ' '); idx > 0 {
			if id, err := strconv.Atoi(s[:idx]); err == nil {
				return id
			}
		}
	}
	return 0
}

// FormatException formats an error with stack trace for the exception field.
func FormatException(err error) string {
	if err == nil {
		return ""
	}

	var buf bytes.Buffer
	buf.WriteString(err.Error())
	buf.WriteString("\n\nStack trace:\n")

	// Get stack trace
	var stack [32]uintptr
	n := runtime.Callers(2, stack[:])
	frames := runtime.CallersFrames(stack[:n])

	for {
		frame, more := frames.Next()
		buf.WriteString("  ")
		buf.WriteString(frame.Function)
		buf.WriteString("\n    ")
		buf.WriteString(frame.File)
		buf.WriteString(":")
		buf.WriteString(strconv.Itoa(frame.Line))
		buf.WriteString("\n")
		if !more {
			break
		}
	}

	return buf.String()
}

File: node/src/formatter.ts
import type { LogContext, LogDocument, LogSource, ConsoleMethod } from './types';

/**
 * Normalize console method name to standard log level.
 */
function normalizeLevel(method: ConsoleMethod): string {
  if (method === 'warn') return 'warning';
  if (method === 'log') return 'info';
  return method;
}

/**
 * Format console arguments into a single message string.
 */
function formatMessage(args: unknown[]): string {
  return args
    .map((arg) => {
      if (typeof arg === 'string') return arg;
      if (arg instanceof Error) return `${arg.name}: ${arg.message}`;
      try {
        return JSON.stringify(arg);
      } catch {
        return String(arg);
      }
    })
    .join(' ');
}

/**
 * Extract source location from an Error stack trace.
 * Walks the stack to find the first frame outside of devlogs internals.
 */
export function extractSourceLocation(): LogSource {
  const base: LogSource = {
    logger: 'node',
    pathname: null,
    lineno: null,
    funcName: null,
  };

  const err = new Error();
  const stack = err.stack;
  if (!stack) return base;

  const lines = stack.split('\n');

  // Skip frames inside devlogs (interceptor, formatter, index)
  for (let i = 1; i < lines.length; i++) {
    const line = lines[i].trim();

    // Skip internal devlogs package frames
    if (
      line.includes('devlogs.cjs') ||
      line.includes('devlogs.esm') ||
      line.includes('node_modules/devlogs-node') ||
      line.includes('/node/dist/') ||
      line.includes('/node/src/interceptor.') ||
      line.includes('/node/src/formatter.') ||
      line.includes('/node/src/index.')
    ) {
      continue;
    }

    // Parse V8 stack frame: "at funcName (file:line:col)" or "at file:line:col"
    const withFunc = line.match(/at\s+(.+?)\s+\((.+?):(\d+):\d+\)/);
    if (withFunc) {
      return {
        logger: 'node',
        pathname: withFunc[2],
        lineno: parseInt(withFunc[3], 10),
        funcName: withFunc[1],
      };
    }

    const noFunc = line.match(/at\s+(.+?):(\d+):\d+/);
    if (noFunc) {
      return {
        logger: 'node',
        pathname: noFunc[1],
        lineno: parseInt(noFunc[2], 10),
        funcName: null,
      };
    }
  }

  return base;
}

/**
 * Extract fields from console arguments if an object is provided.
 */
function extractFields(
  args: unknown[],
  contextFields: Record<string, unknown>
): Record<string, unknown> {
  const fields: Record<string, unknown> = { ...contextFields };

  // If last argument is a plain object, merge it as fields
  const lastArg = args[args.length - 1];
  if (
    lastArg &&
    typeof lastArg === 'object' &&
    !Array.isArray(lastArg) &&
    !(lastArg instanceof Error)
  ) {
    Object.assign(fields, lastArg);
  }

  return fields;
}

/**
 * Format a log entry into the devlogs v2.0 document schema.
 */
export function formatLogDocument(
  method: ConsoleMethod,
  args: unknown[],
  context: LogContext,
  source?: LogSource
): LogDocument {
  const fields = extractFields(args, context.fields);
  const resolvedSource = source ?? extractSourceLocation();

  const doc: LogDocument = {
    doc_type: 'log_entry',
    application: context.application,
    component: context.component,
    timestamp: new Date().toISOString(),
    message: formatMessage(args),
    level: normalizeLevel(method),
    area: context.area,
    operation_id: context.operationId,
    source: resolvedSource,
    process: {
      id: process.pid,
      thread: null,
    },
  };

  if (context.environment) {
    doc.environment = context.environment;
  }
  if (context.version) {
    doc.version = context.version;
  }
  if (Object.keys(fields).length > 0) {
    doc.fields = fields;
  }

  return doc;
}
